src/ngraph_builder.cc

```diff
@@ -361,6 +361,120 @@ tf::Status Builder::TranslateGraph(const std::vector<tf::TensorShape>& inputs,
+
+      if(ng_lhs_shape.size() != ng_rhs_shape.size()) {
+        return tf::errors::InvalidArgument(
+            "Dimensions of two input args are not the same for BatchMatMul");
+      }
+      size_t n_dims = ng_lhs_shape.size();
+      if(n_dims < 2) {
+        return tf::errors::InvalidArgument(
+            "Dimensions of input args for BatchMatMul must be >=2", n_dims);
+      }
+
+      ng::AxisVector out_axes;
+      for (size_t i = 0; i < n_dims - 2; ++i) {
+        if(ng_lhs_shape[i] != ng_rhs_shape[i]){
+          return tf::errors::InvalidArgument(
+              "ng_lhs_shape and ng_rhs_shape must be the same for BatchMatMul for each dimension",
+               i);
+        }
+        out_axes.push_back(i);
+      }
+
+      bool tf_adj_x = false;
+      bool tf_adj_y = false;
+      TF_RETURN_IF_ERROR(tf::GetNodeAttr(op->attrs(), "adj_x", &tf_adj_x));
+      TF_RETURN_IF_ERROR(tf::GetNodeAttr(op->attrs(), "adj_y", &tf_adj_y));
+      
+      auto ng_lhs_axes = out_axes;
+      auto ng_rhs_axes = out_axes;
+      if (tf_adj_x) {
+        ng_lhs_axes.push_back(n_dims-1);
+        ng_lhs_axes.push_back(n_dims-2);
+        ng_lhs = ng::builder::numpy_transpose(ng_lhs, ng_lhs_axes);
+      }
+      if (tf_adj_y) {
+        ng_rhs_axes.insert(ng_rhs_axes.begin(), n_dims-2); 
+        ng_rhs_axes.insert(ng_rhs_axes.begin(), n_dims-1);
+        ng_rhs = ng::builder::numpy_transpose(ng_rhs, ng_rhs_axes);
+      } else {
+        ng_rhs_axes.insert(ng_rhs_axes.begin(), n_dims-1);      
+        ng_rhs_axes.insert(ng_rhs_axes.begin(), n_dims-2);
+        ng_rhs = ng::builder::numpy_transpose(ng_rhs, ng_rhs_axes);
+      }
+
+      ng_lhs_shape = ng_lhs->get_shape();
+      ng_rhs_shape = ng_rhs->get_shape();
+ 
+      if(ng_lhs_shape[n_dims-1] != ng_rhs_shape[0]) {
+        return tf::errors::InvalidArgument(
+            "The last dimension of ng_lhs and the first dimension of ng_rhs should have the same size"
+            );
+      }
+      if (n_dims == 2) {
+        ng_op_map[op->name()] = make_shared<ngraph::op::Dot>(ng_lhs, ng_rhs);
+      } else {
+        auto output_type = ng_lhs->get_element_type(); 
+        auto output_shape = ng_lhs_shape;
+        output_shape[n_dims-1] = ng_rhs_shape[1];
+        auto output_tensor = make_shared<ngraph::op::Parameter>(output_type, output_shape);
+        auto dot_output = make_shared<ngraph::op::Dot>(ng_lhs, ng_rhs);
+        size_t compound_size = 1;
+        for (int i=0; i<out_axes.size(); i++) {
+          compound_size *= output_shape[i];
+        }
+        auto dot_axes = out_axes;
+        dot_axes.push_back(n_dims-2);
+        dot_axes.push_back(n_dims-1);
+        for (int i=0; i<out_axes.size(); i++) {
+          dot_axes.push_back(n_dims+i);
+        }
+        ng::Shape dot_shape = {compound_size, ng_lhs_shape[n_dims-2], ng_rhs_shape[1], compound_size};
+        auto dot_reshape = make_shared<ngraph::op::Reshape>(dot_output, dot_axes, dot_shape);
```

Comment:
This reshape should be conditional on `n_dims > 3`.

Reply:
Added the condition for `dot_reshape` only applies for `n_dims > 3` case